ouroboros-consensus-cardano-test/test/Test/Consensus/Cardano/LedgerTables.hs

```diff
@@ -0,0 +1,90 @@
+import           Ouroboros.Consensus.Cardano.CanHardFork
+import           Ouroboros.Consensus.HardFork.Combinator
+import           Ouroboros.Consensus.HardFork.Combinator.State.Types
+import           Ouroboros.Consensus.HardFork.Combinator.Util.Functors
+import           Ouroboros.Consensus.HardFork.Combinator.Util.Telescope
+import           Ouroboros.Consensus.Ledger.Tables
+import           Ouroboros.Consensus.Protocol.Praos.Translate ()
+import           Ouroboros.Consensus.Shelley.HFEras ()
+import           Ouroboros.Consensus.Shelley.Ledger.SupportsProtocol ()
+import           Ouroboros.Consensus.Shelley.ShelleyHFC (ShelleyTxOut (..))
+import           Ouroboros.Consensus.Util.SOP
+
+import qualified Cardano.Ledger.Shelley.API as SL
+
+import           Test.Cardano.Ledger.Alonzo.Serialisation.Generators ()
+import           Test.Cardano.Ledger.Babbage.Serialisation.Generators ()
+
+import           Test.LedgerTables
+
+import           Test.Consensus.Byron.Generators ()
+import           Test.Consensus.Cardano.MockCrypto (MockCryptoCompatByron)
+import           Test.Consensus.Shelley.Generators ()
+
+import           Test.Tasty
+import           Test.Tasty.QuickCheck
+
+type Crypto = MockCryptoCompatByron
+
+tests :: TestTree
+tests = testGroup "LedgerTables"
+  [ testGroup "Cardano"
+    [ testProperty "Stowable laws" (prop_stowable_laws @(CardanoBlock Crypto))
+    , testProperty "TableStuff laws"
+        (\(InSameEra ls tbs) -> prop_tablestuff_laws @(CardanoBlock Crypto) ls tbs)
+    ]
+  ]
+
+instance Arbitrary (LedgerTables (LedgerState (CardanoBlock Crypto)) ValuesMK) where
+  arbitrary = projectLedgerTables . unstowLedgerTables <$> arbitrary
+
+instance Show (LedgerTables (LedgerState (CardanoBlock Crypto)) EmptyMK) where
+  show x = showsLedgerState x ""
+
+instance Show (LedgerTables (LedgerState (CardanoBlock Crypto)) ValuesMK) where
+  show x = showsLedgerState x ""
+
+-- | We need to generate Ledger states and tables in the same era or otherwise
+-- it wouldn't make sense to try to attach them to a hard fork ledger state,
+-- this data definition does just that.
+data InSameEra =
+  InSameEra
+    !(LedgerState (CardanoBlock Crypto) EmptyMK)
+    !(LedgerTables (LedgerState (CardanoBlock Crypto)) ValuesMK)
+  deriving Show
+
+instance Arbitrary InSameEra where
+  arbitrary = do
+    ls <- arbitrary
+    tbs <- arbitrary `suchThat` (sameEra (getHardForkState $ hardForkLedgerStatePerEra ls) . cardanoUTxOTable)
+    pure $ InSameEra ls tbs
+
+-- | This function asserts that:
+--
+-- - Producing a table to be attached to a Byron HardFork State requires the
+--   table to be empty as that is the only kind of table such a state can have.
+--
+-- - Producing a table to be attached to a Shelley HardFork State requires all
+-- - the entries in the map to be on the same era as the state. We do so by
+-- - checking that all of them and the state produce the same index through
+-- - @'nsToIndex'@.
+sameEra ::
```

Comment:
Hmmm, we have a run-time check for this. Can't this function be defined at the type level?

Reply:
I don't think we can. We cannot introspect into which index an NS is at, i.e. for this:
```
Z (I 'x')      :: NS I       '[ Char, Bool ]
S (Z (I True)) :: NS I       '[ Char, Bool ]
```
they both have the same type, so we have to pattern match on the value to find out the index.

We could leverage the `NS` to carry an index but I think that would trash away the NS purpose and the `nsToIndex` is intended to be used precisely for this